cirq-core/cirq/ops/control_values.py

```diff
@@ -0,0 +1,187 @@
+
+        Returns:
+          An object that represents the cartesian product of the two inputs.
+        """
+
+    @abc.abstractmethod
+    def _expand(self) -> Iterator[Tuple[int, ...]]:
+        """Returns a plain sum of product representation of the values instead
+        of the (possibly compressed) internal representation.
+        """
+
+    @abc.abstractmethod
+    def diagram_repr(self) -> str:
+        """Returns a string representation to be used in circuit diagrams."""
+
+    @abc.abstractmethod
+    def _number_variables(self) -> int:
+        """Returns the number of variables controlled by the object."""
+
+    @abc.abstractmethod
+    def __len__(self) -> int:
+        pass
+
+    @abc.abstractmethod
+    def _identifier(self) -> Any:
+        """Returns the internal representation of the object."""
+
+    @abc.abstractmethod
+    def __hash__(self) -> int:
+        pass
+
+    @abc.abstractmethod
+    def __repr__(self) -> str:
+        pass
+
+    @abc.abstractmethod
+    def validate(self, qid_shapes: Union[Tuple[int, ...], List[int]]) -> None:
+        """Validates control values
+
+        Validate that control values are in the half closed interval
+        [0, qid_shapes) for each qubit.
+        """
+
+    @abc.abstractmethod
+    def _are_ones(self) -> bool:
+        """Checks whether all control values are equal to 1."""
+
+    @abc.abstractmethod
+    def _json_dict_(self) -> Dict[str, Any]:
+        pass
+
+    @abc.abstractmethod
+    def __getitem__(
+        self, key: Union[slice, int]
+    ) -> Union['AbstractControlValues', Tuple[int, ...]]:
+        pass
+
+    def __iter__(self) -> Generator[Tuple[int, ...], None, None]:
+        for assignment in self._expand():
+            assignment = cast(Tuple[int, ...], assignment)
+            yield assignment
+
+    def __eq__(self, other) -> bool:
+        """Returns True iff self and other represent the same configurations.
+
+        Args:
+            other: A AbstractControlValues object.
+
+        Returns:
+            boolean whether the two objects are equivalent or not.
+        """
+        if not isinstance(other, AbstractControlValues):
+            other = ProductOfSums(other)
+        return sorted(v for v in self) == sorted(v for v in other)
+
+
+@dataclass(frozen=True, eq=False)
+class ProductOfSums(AbstractControlValues):
+    """ProductOfSums represents control values in a form of a cartesian product of tuples."""
+
+    _internal_representation: Tuple[Tuple[int, ...], ...]
+
+    def _identifier(self) -> Tuple[Tuple[int, ...], ...]:
+        return self._internal_representation
+
+    def _expand(self) -> Iterator[Tuple[int, ...]]:
+        """Returns the combinations tracked by the object."""
+        self = cast('ProductOfSums', self)
+        return itertools.product(*self._internal_representation)
+
+    def __repr__(self) -> str:
+        return f'cirq.ProductOfSums({str(self._identifier())})'
```

Comment:
Add a test to `control_values_test.py` which asserts equivalent repr using `cirq.testing.assert_equivalent_repr`